lib/server.ml

```diff
@@ -36,70 +36,115 @@ let make channel =
         | Option.Abort -> fail (Failure "client requested abort")
-        | Option.Unknown x ->
-          OptionResponseHeader.(marshal res { request_type = Option.Unknown x; response_type = OptionResponse.Unsupported; length = 0l });
-          channel.write res
+        | Option.Unknown _ as opt ->
+          respond opt OptionResponse.Unsupported chan.write
           >>= fun () ->
           loop ()
-        | Option.List ->
+        | Option.List as opt ->
           begin match offer with
             | None ->
-              OptionResponseHeader.(marshal res { request_type = Option.List; response_type = OptionResponse.Policy; length = 0l });
-              channel.write res
+              respond opt OptionResponse.Policy chan.write
               >>= fun () ->
               loop ()
             | Some offers ->
               let rec advertise = function
-                | [] ->
-                  OptionResponseHeader.(marshal res { request_type = Option.List; response_type = OptionResponse.Ack; length = 0l });
-                  channel.write res
+                | [] -> send_ack opt chan.write
                 | x :: xs ->
                   let l = String.length x in
-                  OptionResponseHeader.(marshal res { request_type = Option.List; response_type = OptionResponse.Server; length = Int32.of_int l});
-                  channel.write res
+                  OptionResponseHeader.(marshal res { request_type = opt; response_type = OptionResponse.Server; length = Int32.of_int l});
+                  chan.write res
                   >>= fun () ->
                   let name = Cstruct.create l in
                   Cstruct.blit_from_string x 0 name 0 l;
-                  channel.write name
+                  chan.write name
                   >>= fun () ->
                   advertise xs in
               advertise offers
               >>= fun () ->
               loop ()
           end
-      end in
-  loop ()
+      end
+    in loop ()
+  in
+  let rec negotiate_tls () =
+    channel.read_clear req
+    >>= fun () ->
+    match OptionRequestHeader.unmarshal req with
+    | Error e -> fail e
+    | Ok hdr ->
+      let payload = make_blank_payload hdr in
+      channel.read_clear payload
+      >>= fun () ->
+      begin match hdr.OptionRequestHeader.ty with
+        | Option.ExportName -> fail (Failure "Client requested export over cleartext channel but server is in FORCEDTLS mode.")
+        | Option.Abort -> fail (Failure "Client requested abort (before negotiating TLS).")
+        | Option.StartTLS as opt -> (
+            send_ack opt channel.write_clear
+            >>= channel.make_tls_channel
+            >>= fun tch ->
+            generic_loop (Channel.generic_of_tls_channel tch)
+          )
+        (* For any other option, we just respond saying TLS is required. *)
+        | opt -> respond opt OptionResponse.TlsReqd channel.write_clear
```

Comment:
We do have to read the payload of all the other request types, to be able to read subsequent requests in this loop, as specified by the NBD protocol.